src/aliceVision/sfmData/colorize.cpp

```diff
@@ -31,97 +62,49 @@ void colorizeTracks(SfMData& sfmData)
   for(auto& landmarkPair : sfmData.getLandmarks())
     remainingLandmarksToColor.push_back(landmarkPair.second);
 
-  struct ViewInfo
-  {
-    ViewInfo(IndexT viewId, std::size_t cardinal)
-      : viewId(viewId)
-      , cardinal(cardinal)
-    {}
-
-    IndexT viewId;
-    std::size_t cardinal;
-    std::vector<std::reference_wrapper<Landmark>> landmarks;
-  };
-
-  std::vector<ViewInfo> sortedViewsCardinal;
-  sortedViewsCardinal.reserve(sfmData.getViews().size());
-  {
-    // create cardinal per viewId map
-    std::map<IndexT, std::size_t> viewsCardinalMap; // <ViewId, Cardinal>
-    for(const auto& landmarkPair : sfmData.getLandmarks())
-    {
-      const Observations& observations = landmarkPair.second.observations;
-      for(const auto& observationPair : observations)
-        ++viewsCardinalMap[observationPair.first]; // TODO: 0
-    }
-
-    // copy key-value pairs from the map to the vector
-    for(const auto& cardinalPair : viewsCardinalMap)
-      sortedViewsCardinal.push_back(ViewInfo(cardinalPair.first, cardinalPair.second));
-
-    // sort the vector, biggest cardinality first
-    std::sort(sortedViewsCardinal.begin(),
-              sortedViewsCardinal.end(),
-              [] (const ViewInfo& l, const ViewInfo& r) { return l.cardinal > r.cardinal; });
-  }
+  std::vector<LMColorAccumulator> landmarkInfo(sfmData.getLandmarks().size());
+  const Views& views = sfmData.getViews();
 
-  // assign each landmark to a view
-  for(ViewInfo& viewCardinal : sortedViewsCardinal)
+  #pragma omp parallel for
+  for( int viewId = 0; viewId < views.size(); ++viewId)
   {
-    std::vector<std::reference_wrapper<Landmark>> toKeep;
-    const IndexT viewId = viewCardinal.viewId;
+      image::Image<image::RGBColor> image;
+      image::readImage(views.at(viewId)->getImagePath(), image, image::EImageColorSpace::SRGB);
 
-    for(int i = 0; i < remainingLandmarksToColor.size(); ++i)
-    {
-      Landmark& landmark = remainingLandmarksToColor.at(i);
-      auto it = landmark.observations.find(viewId);
-      if(it != landmark.observations.end())
-      {
-        viewCardinal.landmarks.push_back(landmark);
-      }
-      else
+      #pragma omp parallel for
+      for(int i = 0; i < remainingLandmarksToColor.size(); ++i)
       {
-        toKeep.push_back(landmark);
+            const Landmark& landmark = remainingLandmarksToColor.at(i);
+            const auto it = landmark.observations.find(viewId);
+
+            if(it != landmark.observations.end())
+            {
+                const Vec3& Tcenter = sfmData.getAbsolutePose(viewId).getTransform().center();
+                const Vec3& pt = landmark.X;
+                const double eucd = 1.0 / (Tcenter - pt).norm();
```

Comment:
The scale of the reconstruction is undefined and the result of this weighting will change with the scale of the scene, isn't it?
Maybe you could precompute the distance of the closest observation per landmark to use it in the normalization of the weighting.

Reply:
No, scale should not matter as long as the entire scene has one scale. The positions of the viewpoints and landmarks should scale together, so the distances of viewpoints to landmark relative to each other will remain the same. So normalizing the weights by their sum, to take the weighted average, is sufficient. I assume that scale of different parts of the same scene do not change individually, if they do please let me know.

I did however notice that I'm still using inverse of the distance, which was intended for debugging visualization purposes. Will remove the inverse to be efficient.